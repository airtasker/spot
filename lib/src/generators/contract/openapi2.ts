import assertNever from "assert-never";
import YAML from "js-yaml";
import compact from "lodash/compact";
import pickBy from "lodash/pickBy";
import {
  BodyDefinition,
  ContractDefinition,
  DefaultResponseDefinition,
  EndpointDefinition
} from "../../models/definitions";
import {
  OpenAPI2SchemaType,
  OpenAPI2SchemaTypeObject,
  openApi2TypeSchema
} from "./openapi2-schema";

const SECURITY_HEADER_SCHEME_NAME = "securityHeader";

export function generateOpenApiV2(
  contractDefinition: ContractDefinition,
  format: "json" | "yaml"
) {
  const contract = openApiV2(contractDefinition);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract, {
        skipInvalid: true // for undefined
      });
    default:
      throw assertNever(format);
  }
}

export function openApiV2(contractDefinition: ContractDefinition): OpenApiV2 {
  return {
    swagger: "2.0",
    info: {
      version: "0.0.0",
      title: contractDefinition.api.name,
      ...pickBy({ description: contractDefinition.api.description }),
      contact: {
        name: "TODO"
      }
    },
    paths: contractDefinition.endpoints.reduce<{
      [endpointPath: string]: {
        [method: string]: OpenAPIV2Operation;
      };
    }>((acc, endpoint) => {
      const openApiPath = endpoint.path.replace(/:(\w+)/g, "{$1}");
      acc[openApiPath] = acc[openApiPath] || {};
      acc[openApiPath][endpoint.method.toLowerCase()] = {
        operationId: endpoint.name,
        description: endpoint.description,
        tags: endpoint.tags,
        parameters: getParameters(endpoint),
        responses: {
          ...(endpoint.defaultResponse
            ? { default: response(endpoint.defaultResponse) }
            : {}),
          ...endpoint.responses.reduce<{
            [statusCode: string]: OpenAPIV2Response;
          }>((acc, responseNode) => {
            acc[responseNode.status.toString(10)] = response(responseNode);
            return acc;
          }, {})
        }
      };
      return acc;
    }, {}),
    definitions: contractDefinition.types.reduce<{
      [typeName: string]: OpenAPI2SchemaType;
    }>((acc, typeNode) => {
      acc[typeNode.name] = openApi2TypeSchema(typeNode.type);
      return acc;
    }, {}),
    ...(contractDefinition.api.securityHeader
      ? {
          securityDefinitions: {
            [SECURITY_HEADER_SCHEME_NAME]: {
              type: "apiKey",
              in: "header",
              name: contractDefinition.api.securityHeader.name,
              description: contractDefinition.api.securityHeader.description
            }
          },
          security: [
            {
              [SECURITY_HEADER_SCHEME_NAME]: []
            }
          ]
        }
      : {})
  };
}

function getParameters(endpoint: EndpointDefinition): OpenAPIV2Parameter[] {
  const parameters = endpoint.request.pathParams
    .map(
      (pathParam): OpenAPIV2Parameter => {
        const schemaType = openApi2TypeSchema(pathParam.type);
        if ("type" in schemaType && schemaType.type === "object") {
          throw new Error(`Unsupported object type in path parameter`);
        }
        return {
          in: "path",
          name: pathParam.name,
          description: pathParam.description,
          ...schemaType,
          required: true
        };
      }
    )
    .concat(endpoint.request.body ? [requestBody(endpoint.request.body)] : [])
    .concat(
      endpoint.request.queryParams.map(
        (queryParam): OpenAPIV2Parameter => {
          const schemaType = openApi2TypeSchema(queryParam.type);
          if ("type" in schemaType && schemaType.type === "object") {
            throw new Error(`Unsupported object type in query parameter`);
          }
          return {
            in: "query",
            name: queryParam.name,
            description: queryParam.description,
            ...schemaType,
            required: !queryParam.optional
          };
        }
      )
    )
    .concat(
      endpoint.request.headers.map(
        (header): OpenAPIV2Parameter => {
          const schemaType = openApi2TypeSchema(header.type);
          if ("type" in schemaType && schemaType.type === "object") {
            throw new Error(`Unsupported object type in header`);
          }
          return {
            in: "header",
            name: header.name,
            description: header.description,
            ...schemaType,
            required: !header.optional
          };
        }
      )
    );
  return compact(parameters);
}

function requestBody(body: BodyDefinition): OpenAPIV2Parameter {
  return {
    in: "body",
    name: "body",
    description: "TODO",
    required: true,
    schema: openApi2TypeSchema(body.type)
  };
}

function response(response: DefaultResponseDefinition): OpenAPIV2Response {
  return {
    schema: response.body && openApi2TypeSchema(response.body.type),
    headers: response.headers.reduce<{
      [header: string]: OpenAPIV2Header;
    }>((headerAcc, header) => {
      const schemaType = openApi2TypeSchema(header.type);
      headerAcc[header.name] = {
        ...schemaType,
        description: header.description
      };
      return headerAcc;
    }, {}),
    description: response.description || ""
  };
}

export interface OpenApiV2 {
  swagger: "2.0";
  tags?: OpenAPIV2TagObject[];
  info: {
    version: string;
    title: string;
    description?: string;
    termsOfService?: string;
    contact?: {
      name?: string;
      url?: string;
      email?: string;
    };
    license?: {
      name: string;
      url?: string;
    };
  };
  host?: string;
  paths: {
    [endpointPath: string]: {
      [method: string]: OpenAPIV2Operation;
    };
  };
  definitions: {
    [typeName: string]: OpenAPI2SchemaType;
  };
  securityDefinitions?: {
    [securitySchemeName: string]: OpenAPIV2SecurityScheme;
  };
  security?: {
    [securitySchemeName: string]: string[];
  }[];
}

export interface OpenAPIV2TagObject {
  name: string;
  description?: string;
}

export interface OpenAPIV2Operation {
  operationId: string;
  description?: string;
  tags?: string[];
  parameters: OpenAPIV2Parameter[];
  requestBody?: OpenAPI2SchemaType;
  responses: {
    default?: OpenAPIV2Response;
    // Note: we use | undefined because otherwise "default" would have to be required.
    [statusCode: string]: OpenAPIV2Response | undefined;
  };
}

export type OpenAPIV2Parameter =
  | OpenAPIV2BodyParameter
  | OpenAPIV2NonBodyParameter;

export type OpenAPIV2NonBodyParameter = {
  in: "path" | "query" | "header";
  name: string;
  description?: string;
  required: boolean;
} & Exclude<OpenAPI2SchemaType, OpenAPI2SchemaTypeObject>;

export type OpenAPIV2BodyParameter = {
  in: "body";
  name: "body";
  description?: string;
  required: boolean;
  schema: OpenAPI2SchemaType | undefined;
};

export interface OpenAPIV2Response {
  schema?: OpenAPI2SchemaType;
  headers?: {
    [header: string]: OpenAPIV2Header;
  };
  description: string;
}

export type OpenAPIV2Header = {
  description?: string;
} & OpenAPI2SchemaType;

// TODO: Consider adding support for other security schemes.
export type OpenAPIV2SecurityScheme = OpenApiV2SecurityScheme_ApiKey;

export interface OpenApiV2SecurityScheme_ApiKey {
  type: "apiKey";
  description?: string;
  name: string;
  in: "query" | "header";
}
