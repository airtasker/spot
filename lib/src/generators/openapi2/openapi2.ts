import assertNever from "assert-never";
import {
  Config,
  Contract,
  DefaultResponse,
  Endpoint,
  HttpMethod,
  isSpecificResponse,
  Request,
  Response
} from "../../definitions";
import { TypeTable } from "../../types";
import { KeyOfType } from "../../util";
import {
  pathParamToPathParameterObject,
  queryParamToQueryParameterObject,
  requestHeaderToHeaderParameterObject,
  responseHeaderToHeaderObject
} from "./openapi2-parameter-util";
import {
  BodyParameterObject,
  DefinitionsObject,
  HeaderObject,
  HeaderParameterObject,
  OpenApiV2,
  OperationObject,
  ParameterObject,
  PathItemObject,
  PathParameterObject,
  PathsObject,
  QueryParameterObject,
  ResponseObject,
  ResponsesObject,
  SecurityDefinitionsObject
} from "./openapi2-specification";
import { typeToSchemaObject } from "./openapi2-type-util";

const SECURITY_HEADER_SCHEME_NAME = "SecurityHeader";

export function generateOpenAPI2(contract: Contract): OpenApiV2 {
  const typeTable = TypeTable.fromArray(contract.types);
  const openapi: OpenApiV2 = {
    swagger: "2.0",
    info: {
      title: contract.name,
      description: contract.description,
      version: "0.0.0"
    },
    consumes: ["application/json"],
    produces: ["application/json"],
    paths: endpointsToPathsObject(
      contract.endpoints,
      typeTable,
      contract.config
    ),
    definitions: contractTypesToDefinitionsObject(contract.types, typeTable),
    securityDefinitions: contractSecurityToSecurityDefinitionsObject(
      contract.security
    ),
    security: contract.security && [
      {
        [SECURITY_HEADER_SCHEME_NAME]: []
      }
    ]
  };

  return openapi;
}

function contractTypesToDefinitionsObject(
  types: Contract["types"],
  typeTable: TypeTable
): DefinitionsObject | undefined {
  if (types.length === 0) {
    return;
  }

  return types.reduce<DefinitionsObject>((acc, t) => {
    acc[t.name] = typeToSchemaObject(t.type, typeTable);
    return acc;
  }, {});
}

function contractSecurityToSecurityDefinitionsObject(
  security: Contract["security"]
): SecurityDefinitionsObject | undefined {
  return (
    security && {
      [SECURITY_HEADER_SCHEME_NAME]: {
        type: "apiKey",
        in: "header",
        name: security.name,
        description: security.description
      }
    }
  );
}

function endpointsToPathsObject(
  endpoints: Endpoint[],
  typeTable: TypeTable,
  config: Config
): PathsObject {
  return endpoints.reduce<PathsObject>((acc, endpoint) => {
    const pathName = endpoint.path
      .split("/")
      .map(component =>
        component.startsWith(":") ? `{${component.slice(1)}}` : component
      )
      .join("/");

    acc[pathName] = acc[pathName] ?? {};
    const pathItemMethod = httpMethodToPathItemMethod(endpoint.method);
    acc[pathName][pathItemMethod] = endpointToOperationObject(
      endpoint,
      typeTable,
      config
    );
    return acc;
  }, {});
}

function endpointToOperationObject(
  endpoint: Endpoint,
  typeTable: TypeTable,
  config: Config
): OperationObject {
  const endpointRequest = endpoint.request;

  return {
    tags: endpoint.tags.length > 0 ? endpoint.tags : undefined,
    description: endpoint.description,
    operationId: endpoint.name,
    parameters:
      endpointRequest &&
      endpointRequestToParameterObjects(endpointRequest, typeTable, config),
    responses: endpointResponsesToResponsesObject(
      {
        specific: endpoint.responses,
        default: endpoint.defaultResponse
      },
      typeTable
    )
  };
}

function endpointRequestToParameterObjects(
  request: Request,
  typeTable: TypeTable,
  config: Config
): ParameterObject[] {
  const pathParameters: PathParameterObject[] = request.pathParams.map(p =>
    pathParamToPathParameterObject(p, typeTable)
  );

  const queryParameters: QueryParameterObject[] = request.queryParams.map(p =>
    queryParamToQueryParameterObject(p, typeTable, config)
  );

  const headerParameters: HeaderParameterObject[] = request.headers.map(p =>
    requestHeaderToHeaderParameterObject(p, typeTable)
  );

  const bodyParameter: BodyParameterObject | undefined = request.body && {
    in: "body",
    name: "Body", // name has no effect on the body parameter
    required: true, // TODO: currently Spot does not support optional request body
    schema: typeToSchemaObject(request.body.type, typeTable)
  };

  const parameters: ParameterObject[] = [];

  return parameters
    .concat(pathParameters, queryParameters, headerParameters)
    .concat(bodyParameter ?? []);
}

function endpointResponsesToResponsesObject(
  responses: {
    specific: Response[];
    default?: DefaultResponse;
  },
  typeTable: TypeTable
): ResponsesObject {
  const responsesObject = responses.specific.reduce<ResponsesObject>(
    (acc, response) => {
      acc[response.status.toString(10)] = endpointResponseToResponseObject(
        response,
        typeTable
      );
      return acc;
    },
    {}
  );

  if (responses.default) {
    responsesObject.default = endpointResponseToResponseObject(
      responses.default,
      typeTable
    );
  }

  return responsesObject;
}

function endpointResponseToResponseObject(
  response: Response | DefaultResponse,
  typeTable: TypeTable
): ResponseObject {
  const description =
    response.description ??
    (isSpecificResponse(response)
      ? `${response.status} response`
      : "default response");

  const headers =
    response.headers.length > 0
      ? response.headers.reduce<{ [name: string]: HeaderObject }>(
          (acc, header) => {
            acc[header.name] = responseHeaderToHeaderObject(header, typeTable);
            return acc;
          },
          {}
        )
      : undefined;

  const schema =
    response.body && typeToSchemaObject(response.body.type, typeTable);

  return { description, headers, schema };
}

function httpMethodToPathItemMethod(
  method: HttpMethod
): KeyOfType<PathItemObject, OperationObject> {
  switch (method) {
    case "GET":
      return "get";
    case "PUT":
      return "put";
    case "POST":
      return "post";
    case "DELETE":
      return "delete";
    case "PATCH":
      return "patch";
    default:
      assertNever(method);
  }
}
