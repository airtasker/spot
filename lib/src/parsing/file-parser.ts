import * as fs from "fs-extra";
import * as path from "path";
import * as ts from "typescript";
import { Api } from "../models";
import { validate } from "../validator";
import { parseApiClass } from "./api-class-parser";
import { extractSingleDecorator } from "./decorators";
import { panic } from "./panic";
import { extractObjectType, extractType } from "./type-parser";

/**
 * Parses a TypeScript source file, as well as any other TypeScript files it imports recursively.
 *
 * This should be smart enough not to get stuck into an infinite loop for circular dependencies.
 */
export async function parsePath(sourcePath: string): Promise<Api> {
  const api: Api = {
    endpoints: {},
    types: {}
  };
  await parseFileRecursively(api, new Set(), sourcePath);
  const errors = validate(api);
  if (errors.length > 0) {
    throw panic(errors.join("\n"));
  }
  return api;
}

async function parseFileRecursively(
  api: Api,
  visitedPaths: Set<string>,
  sourcePath: string
): Promise<void> {
  if (!(await fs.existsSync(sourcePath))) {
    if (await fs.existsSync(sourcePath + ".ts")) {
      sourcePath += ".ts";
    } else {
      throw panic(`No source file found at ${sourcePath}`);
    }
  }
  if (
    visitedPaths.has(sourcePath) ||
    path.resolve(sourcePath).startsWith(__dirname)
  ) {
    return;
  } else {
    visitedPaths.add(sourcePath);
  }
  const fileContent = await fs.readFile(sourcePath, "utf8");
  const sourceFile = ts.createSourceFile(
    path.basename(sourcePath),
    fileContent,
    ts.ScriptTarget.Latest
  );
  for (const statement of sourceFile.statements) {
    if (ts.isImportDeclaration(statement)) {
      if (!ts.isStringLiteral(statement.moduleSpecifier)) {
        throw panic(
          `Unsupported import statement: ${statement.moduleSpecifier.getText(
            sourceFile
          )}`
        );
      }
      const importPath = statement.moduleSpecifier.text;
      if (!importPath.startsWith(".")) {
        // This is not a relative import, we'll ignore it.
        continue;
      }
      await parseFileRecursively(
        api,
        visitedPaths,
        path.join(sourcePath, "..", importPath)
      );
    } else if (ts.isClassDeclaration(statement)) {
      const apiDecorator = extractSingleDecorator(sourceFile, statement, "api");
      if (apiDecorator) {
        parseApiClass(sourceFile, statement, api);
      }
    } else if (ts.isTypeAliasDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractType(sourceFile, statement.type);
    } else if (ts.isInterfaceDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractObjectType(sourceFile, statement);
    }
  }
}
