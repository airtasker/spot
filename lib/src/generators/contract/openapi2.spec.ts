import * as fs from "fs-extra";
import * as path from "path";
import { parsePath } from "../../parsing/file-parser";
import { generateOpenApiV2 } from "./openapi2";

const EXAMPLES_DIR = path.join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "examples",
  "src"
);

describe("OpenAPI 2 generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, "api.ts")
        );
        expect(generateOpenApiV2(api, "json")).toMatchSnapshot("json");
        expect(generateOpenApiV2(api, "yaml")).toMatchSnapshot("yaml");
      });
    }
  });
});
