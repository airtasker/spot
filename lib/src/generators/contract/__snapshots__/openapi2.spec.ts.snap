// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI 2 generator produces valid code multi-file: json 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"tags\\": [
    {
      \\"name\\": \\"TODO\\"
    }
  ],
  \\"info\\": {
    \\"version\\": \\"0.0.0\\",
    \\"title\\": \\"TODO\\",
    \\"contact\\": {
      \\"name\\": \\"TODO\\"
    }
  },
  \\"paths\\": {
    \\"/users/create\\": {
      \\"post\\": {
        \\"operationId\\": \\"createUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [
          \\"application/json\\"
        ],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"body\\",
            \\"description\\": \\"TODO\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/CreateUserRequest\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/CreateUserResponse\\"
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users/{userId}-confirmed\\": {
      \\"delete\\": {
        \\"operationId\\": \\"deleteUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"\\"
          },
          \\"403\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                },
                \\"signedInAs\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\",
                \\"signedInAs\\"
              ]
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\"
              ]
            },
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"findUsers\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"limit\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"number\\",
            \\"required\\": false
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"search_term\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"age\\": {
                    \\"type\\": \\"number\\"
                  }
                },
                \\"required\\": [
                  \\"name\\"
                ]
              }
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users/{userId}\\": {
      \\"get\\": {
        \\"operationId\\": \\"getUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"number\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    }
  },
  \\"definitions\\": {
    \\"CreateUserRequest\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    },
    \\"CreateUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"success\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"success\\"
      ]
    },
    \\"GetUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    }
  }
}"
`;

exports[`OpenAPI 2 generator produces valid code multi-file: yaml 1`] = `
"swagger: '2.0'
tags:
  - name: TODO
info:
  version: 0.0.0
  title: TODO
  contact:
    name: TODO
paths:
  /users/create:
    post:
      operationId: createUser
      description: TODO
      consumes:
        - application/json
      tags:
        - TODO
      parameters:
        - in: body
          name: body
          description: TODO
          required: true
          schema:
            $ref: '#/definitions/CreateUserRequest'
      responses:
        '200':
          schema:
            $ref: '#/definitions/CreateUserResponse'
          description: ''
        default:
          description: ''
  '/users/{userId}-confirmed':
    delete:
      operationId: deleteUser
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: path
          name: userId
          description: TODO
          type: string
          required: true
      responses:
        '200':
          description: ''
        '403':
          schema:
            type: object
            properties:
              message:
                type: string
              signedInAs:
                type: string
            required:
              - message
              - signedInAs
          description: ''
        default:
          schema:
            type: object
            properties:
              message:
                type: string
            required:
              - message
          description: ''
  /users:
    get:
      operationId: findUsers
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: query
          name: limit
          description: TODO
          type: number
          required: false
        - in: query
          name: search_term
          description: TODO
          type: string
          required: true
      responses:
        '200':
          schema:
            type: array
            items:
              type: object
              properties:
                name:
                  type: string
                age:
                  type: number
              required:
                - name
          description: ''
        default:
          description: ''
  '/users/{userId}':
    get:
      operationId: getUser
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: path
          name: userId
          description: TODO
          type: string
          required: true
      responses:
        '200':
          schema:
            type: object
            properties:
              name:
                type: string
              age:
                type: number
            required:
              - name
          description: ''
        default:
          description: ''
definitions:
  CreateUserRequest:
    type: object
    properties:
      name:
        type: string
    required:
      - name
  CreateUserResponse:
    type: object
    properties:
      success:
        type: boolean
    required:
      - success
  GetUserResponse:
    type: object
    properties:
      name:
        type: string
    required:
      - name
"
`;

exports[`OpenAPI 2 generator produces valid code single-file: json 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"tags\\": [
    {
      \\"name\\": \\"TODO\\"
    }
  ],
  \\"info\\": {
    \\"version\\": \\"0.0.0\\",
    \\"title\\": \\"TODO\\",
    \\"contact\\": {
      \\"name\\": \\"TODO\\"
    }
  },
  \\"paths\\": {
    \\"/users/create\\": {
      \\"post\\": {
        \\"operationId\\": \\"createUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [
          \\"application/json\\"
        ],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"body\\",
            \\"description\\": \\"TODO\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/CreateUserRequest\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/CreateUserResponse\\"
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"findUsers\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"limit\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"number\\",
            \\"required\\": false
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"search_term\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"age\\": {
                    \\"type\\": \\"number\\"
                  }
                },
                \\"required\\": [
                  \\"name\\"
                ]
              }
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users/{userId}\\": {
      \\"get\\": {
        \\"operationId\\": \\"getUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                },
                \\"age\\": {
                  \\"type\\": \\"number\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"description\\": \\"\\"
          }
        }
      }
    },
    \\"/users/{userId}-confirmed\\": {
      \\"delete\\": {
        \\"operationId\\": \\"deleteUser\\",
        \\"description\\": \\"TODO\\",
        \\"consumes\\": [],
        \\"tags\\": [
          \\"TODO\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"TODO\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"\\"
          },
          \\"403\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                },
                \\"signedInAs\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\",
                \\"signedInAs\\"
              ]
            },
            \\"description\\": \\"\\"
          },
          \\"default\\": {
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\"
              ]
            },
            \\"description\\": \\"\\"
          }
        }
      }
    }
  },
  \\"definitions\\": {
    \\"CreateUserRequest\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    },
    \\"CreateUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"success\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"success\\"
      ]
    },
    \\"GetUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    }
  }
}"
`;

exports[`OpenAPI 2 generator produces valid code single-file: yaml 1`] = `
"swagger: '2.0'
tags:
  - name: TODO
info:
  version: 0.0.0
  title: TODO
  contact:
    name: TODO
paths:
  /users/create:
    post:
      operationId: createUser
      description: TODO
      consumes:
        - application/json
      tags:
        - TODO
      parameters:
        - in: body
          name: body
          description: TODO
          required: true
          schema:
            $ref: '#/definitions/CreateUserRequest'
      responses:
        '200':
          schema:
            $ref: '#/definitions/CreateUserResponse'
          description: ''
        default:
          description: ''
  /users:
    get:
      operationId: findUsers
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: query
          name: limit
          description: TODO
          type: number
          required: false
        - in: query
          name: search_term
          description: TODO
          type: string
          required: true
      responses:
        '200':
          schema:
            type: array
            items:
              type: object
              properties:
                name:
                  type: string
                age:
                  type: number
              required:
                - name
          description: ''
        default:
          description: ''
  '/users/{userId}':
    get:
      operationId: getUser
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: path
          name: userId
          description: TODO
          type: string
          required: true
      responses:
        '200':
          schema:
            type: object
            properties:
              name:
                type: string
              age:
                type: number
            required:
              - name
          description: ''
        default:
          description: ''
  '/users/{userId}-confirmed':
    delete:
      operationId: deleteUser
      description: TODO
      consumes: []
      tags:
        - TODO
      parameters:
        - in: path
          name: userId
          description: TODO
          type: string
          required: true
      responses:
        '200':
          description: ''
        '403':
          schema:
            type: object
            properties:
              message:
                type: string
              signedInAs:
                type: string
            required:
              - message
              - signedInAs
          description: ''
        default:
          schema:
            type: object
            properties:
              message:
                type: string
            required:
              - message
          description: ''
definitions:
  CreateUserRequest:
    type: object
    properties:
      name:
        type: string
    required:
      - name
  CreateUserResponse:
    type: object
    properties:
      success:
        type: boolean
    required:
      - success
  GetUserResponse:
    type: object
    properties:
      name:
        type: string
    required:
      - name
"
`;
